res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+    var samplerGroup = "[Sampler" + samplerNumber + "]";
+    if (TraktorS2MK2.shift_pressed[group]) {
+      if (engine.getValue(samplerGroup, "play") === 1) {
+        engine.setValue(samplerGroup, "play", 0);
+      } else {
+        script.triggerControl(samplerGroup, "eject");
+      }
+    } else {
+      if (engine.getValue(samplerGroup, "track_loaded") === 0) {
+        script.triggerControl(samplerGroup, "LoadSelectedTrack");
+      } else {
+        script.triggerControl(samplerGroup, "cue_gotoandplay");
+      }
+    }
+  }
+}
+
+TraktorS2MK2.samplerModeButton = function(field) {
+  if (field.value === 0) {
+    return;
+  }
+  var padMode = TraktorS2MK2.current_pad_mode[field.group];
+  if (padMode !== TraktorS2MK2.pad_modes.sampler) {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.sampler);
+    TraktorS2MK2.controller.setOutput(field.group, "!remix_button", 0x7f, false);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x00, !TraktorS2MK2.freeze_lights);
+  } else {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.hotcue);
+    TraktorS2MK2.controller.setOutput(field.group, "!remix_button", 0x00, !TraktorS2MK2.freeze_lights);
+  }
+}
+
+TraktorS2MK2.introOutroModeButton = function(field) {
+  if (field.value === 0) {
+    return;
+  }
+  var padMode = TraktorS2MK2.current_pad_mode[field.group];
+  if (padMode !== TraktorS2MK2.pad_modes.intro_outro) {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.intro_outro);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x7f, false);
+    TraktorS2MK2.controller.setOutput(field.group, "!remix_button", 0x00, !TraktorS2MK2.freeze_lights);
+  } else {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.hotcue);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x00, !TraktorS2MK2.freeze_lights);
+  }
+}
+
+TraktorS2MK2.loopIn = function(field) {
+  var group = field.id.split(".")[0];
+  engine.setValue(group, "loop_in", field.value);
+}
+
+TraktorS2MK2.loopOut = function(field) {
+  var group = field.id.split(".")[0];
+  engine.setValue(group, "loop_out", field.value);
+}
+
+TraktorS2MK2.samplerActivateButton = function(field) {
+  if (field.value > 0) {
+    TraktorS2Mk2.controller.samplers_active[field.group] = !TraktorS2Mk2.controller.samplers_active[field.group];
+  }
+}
+
+TraktorS2MK2.connectEffectButtonLEDs = function(effectUnitGroup) {
+  TraktorS2MK2.effectButtonLEDconnections[effectUnitGroup].forEach(function(connection) {
+    connection.disconnect();
+  });
+  
+  var focusedEffect = engine.getValue(effectUnitGroup, "focused_effect");
+  var makeButtonLEDcallback = function(effectNumber) {
+    return function(value, group, control) {
+      TraktorS2MK2.controller.setOutput(effectUnitGroup, "!effectbutton" + effectNumber,
+        value === 1 ? 0x7f : 0, !TraktorS2MK2.freeze_lights);
+    };
+  };
+  
+  // FIXME: Why do the LEDs flicker?
```

Comment:
sometimes the controller wants to do the lights itself -- that was the case on my VCI400

Reply:
I think there's a subtle issue with the code, but I haven't figured it out.